src/main/groovy/io/jenkins/infra/repository_permissions_updater/ArtifactoryPermissionsUpdater.groovy

```diff
@@ -288,9 +323,58 @@ public class ArtifactoryPermissionsUpdater {
      * @param args unused
      */
     public static void main(String[] args) {
-        generateApiPayloads(DEFINITIONS_DIR, ARTIFACTORY_API_DIR)
-        submitPermissionTargets(ARTIFACTORY_API_DIR)
-        removeExtraPermissionTargets(ARTIFACTORY_API_DIR)
+        if (!System.getProperty('java.util.logging.SimpleFormatter.format')) {
+            System.setProperty('java.util.logging.SimpleFormatter.format',
+                    '%1$tY-%1$tm-%1$td %1$tH:%1$tM:%1$tS %4$s: %5$s%6$s%n');
+        }
+        def defDefnDir = System.getProperty('definitionsDir', './permissions');
+        def defWorkDir = System.getProperty('artifactoryApiTempDir', './json');
+        def cli = new CliBuilder(usage: 'repository-permissions-updater -dhwto')
+        cli.with {
+            h longOpt: 'help', 'Show usage information'
+            d longOpt: 'definitionDir', args:1, argName: 'dir', "Definitions directory (defaults to ${defDefnDir})"
+            w longOpt: 'workDir', args:1, argName: 'dir', "Work directory (defaults to ${defWorkDir})"
+            t longOpt: 'dryRun', 'Enable read only dry run mode'
+            o longOpt: 'offline', 'Off-line mode (implies dry run but does not perform any remote operations)'
+        }
+        def options = cli.parse(args)
+        if (!options) {
+            return
+        }
+        if (options.h) {
+            cli.usage()
+            return
+        }
+        boolean dryRun = options.t
+        boolean offline = false
+        if (!System.getenv("ARTIFACTORY_USERNAME")) {
+            LOGGER.log(Level.WARNING, "Environment variable 'ARTIFACTORY_USERNAME' not configured, forcing --offline")
+            dryRun = true
+            offline = true
+        }
+        if (!System.getenv('ARTIFACTORY_PASSWORD')) {
+            LOGGER.log(Level.WARNING, "Environment variable 'ARTIFACTORY_PASSWORD' not configured, forcing --offline")
+            dryRun = true
+            offline = true
+        }
+        if (!offline) {
+            /* Make sure all Artifactory API requests are properly authenticated */
+            Authenticator.setDefault(new Authenticator() {
+                protected PasswordAuthentication getPasswordAuthentication() {
+                    return new PasswordAuthentication(
+                            System.getenv("ARTIFACTORY_USERNAME"),
+                            System.getenv("ARTIFACTORY_PASSWORD").toCharArray());
+                }
+            });
+        }
+
+        File definitionsDir = new File(options.d?:defDefnDir)
+        File workDir = new File(options.w?:defWorkDir)
+        generateApiPayloads(definitionsDir, workDir)
+        submitPermissionTargets(workDir, dryRun)
+        if (!offline) {
```

Comment:
dry run when on-line will tell you the permissions for removal.

the submit doesn't care about on-line once dryRun is enabled, but the clean up needs auth